### PR TITLE
Add __array_ufunc__ protocol to DataFrame and Series (#3044)

### PR DESCRIPTION
**Pull request by almaleksia**
* Add __array_ufunc__ protocol to dask.dataframe

    Example
    -------

    ```python
    In [1]: import numpy as np, dask.dataframe as dd, pandas as pd
    In [2]: numpy_arr = np.random.randn(5)
    In [3]: ps = pd.Series(numpy_arr)
    In [5]: np.sum(np.ceil(ps))
    Out[5]: 2.0
    In [6]: ds = dd.from_pandas(ps,3)
    In [9]: np.sum(np.ceil(ds))
    Out[5]: dd.Scalar<series-..., dtype=float64>
    In [6]: _.compute()
    Out[6]: 2.0
    ```
  * Add out parameter support

    Example
    -------

    ```python
    In [1]: import numpy as np, dask.dataframe as dd, pandas as pd
    In [2]: ps = pd.Series(np.random.randn(5))
    In [3]: ds = dd.from_pandas(ps, 3)
    In [4]: dsc_out = dd.core.Scalar({('s', 0): 0}, 's', 'i8')
    In [5]: np.sum(ds, out=dsc_out)
    In [6]: dsc_out
    Out[6]: dd.Scalar<series-..., dtype=float64>
    In [7]: dsc_out.compute()
    Out[7]: 0.2768562049182187
    In [8]: np.sum(ps)
    Out[8]: 0.2768562049182187
    ```

- [ ] Tests added / passed
- [ ] Passes `flake8 dask`
- [ ] Fully documented, including `docs/source/changelog.rst` for all changes
      and one of the `docs/source/*-api.rst` files for new API

### CODE REVIEW

**Comment by mrocklin** [comment]
I've only briefly gone through this, but in general this looks excellent to me.  I'm really looking forward to playing around with it. 

cc @TomAugspurger @shoyer you may both find this interesting.  Also cc @sinhrks if he's around.

File: dask/array/ufunc.py [hunk 1]
```diff
@@ -109,9 +109,13 @@ def __repr__(self):
...
-        dsk = [arg for arg in args if hasattr(arg, '_elemwise')]
-        if len(dsk) > 0:
-            return dsk[0]._elemwise(self._ufunc, *args, **kwargs)
+        dsks = [arg for arg in args if hasattr(arg, '_elemwise')]
+        if len(dsks) > 0:
+            for dsk in dsks:
+                result = dsk._elemwise(self._ufunc, *args, **kwargs)
+                if type(result) != type(NotImplemented):
+                    return result
+            return NotImplemented
```
Comment (shoyer) [inline dask/array/ufunc.py]: Calling a ufunc shouldn't ever return NotImplemented. Instead, if every implementation returns `NotImplemented`  you should raise `TypeError` (like how Python handles arithmetic).

File: dask/dataframe/core.py [hunk 2]
```diff
@@ -3107,7 +3140,49 @@ def elemwise(op, *args, **kwargs):
...
+    if isinstance(out, tuple):
+        if len(out) == 1:
+            out = out[0]
+        elif len(out) > 1:
+            raise NotImplementedError("The out parameter is not fully supported")
+        else:
+            out = None
+
+    if out is not None and type(out) != type(result):
+        raise NotImplementedError(
```
Comment (shoyer) [inline dask/dataframe/core.py]: I think this should be `TypeError`?

File: dask/dataframe/core.py [hunk 3]
```diff
@@ -3107,7 +3140,49 @@ def elemwise(op, *args, **kwargs):
...
+            "Mismatched types between result and out parameter. "
+            "out=%s, result=%s" % (str(type(out)), str(type(result))))
+
+    if isinstance(out, DataFrame):
+        if len(out.columns) != len(result.columns):
+            raise NotImplementedError(
```
Comment (shoyer) [inline dask/dataframe/core.py]: `ValueError`?

File: dask/dataframe/tests/test_arithmetics_reduction.py [hunk 4]
```diff
@@ -688,11 +689,51 @@ def test_reductions(split_every):
...
+                          0, pd.Series([])),
+                          (pd.DataFrame({'a': [1, 2, 3],
+                                        'b': [4, 5, 6]},
+                                        index=[0, 1, 3]),
+                           1, pd.Series([])),
+                          (pd.Series([1, 2.5, 6]), None, None)])
+@pytest.mark.parametrize('redfunc', ['sum', 'prod', 'min', 'max', 'mean', 'var', 'std'])
+def test_reductions_out(frame, axis, out, redfunc):
+    dsk_in = dd.from_pandas(frame, 3)
+    dsk_out = dd.core.Scalar({('a', 0): np.int64(0)}, 'a', 'i8')
```
Comment (mrocklin) [inline dask/dataframe/tests/test_arithmetics_reduction.py]: I recommend avoiding creating Dask objects by hand.  This API is less reliable than the user/pandas API.  Instead I recommend creating a scalar using Pandas API on a dask dataframe:

    dsk_in = dd.from_pandas(frame, 3)
    dsk_out = dsk_in.a.sum()

File: dask/dataframe/core.py (hunk 3 above)
Comment (mrocklin) [reply dask/dataframe/core.py]: I think that this should probably raise a `ValueError` in both cases.  My apologies for the earlier implementation.

**Comment by mrocklin** [comment]
@shoyer thank you for your attentive reivew.  It looks like the issues that you raised have been addressed. 
 Do you have any further comments on this?

File: dask/dataframe/core.py [hunk 5]
```diff
@@ -340,6 +341,26 @@ def __array__(self, dtype=None, **kwargs):
     def __array_wrap__(self, array, context=None):
         raise NotImplementedError
 
+    def __array_ufunc__(self, numpy_ufunc, method, *inputs, **kwargs):
+        out = kwargs.get('out', ())
+        for x in inputs + out:
+            if not isinstance(x, (Number, Scalar, _Frame, Array,
+                                  pd.DataFrame, pd.Series, pd.Index)):
```
Comment (shoyer) [inline dask/dataframe/core.py]: Maybe include `np.ndarray` in this list?

Comment (almaleksia) [reply dask/dataframe/core.py]: @shoyer sure. Currently ndarray is not handled as expected in ufunc wrapper for dataframes. I will need to dig a bit more into it to get it work.

Comment (mrocklin) [reply dask/dataframe/core.py]: We could also merge this now and handle np.ndarray in future work.  The decision for this is probably up to you @almaleksia .

Comment (mrocklin) [reply dask/dataframe/core.py]: OK great, lets leave that for future work then

**Comment by mrocklin** [comment]
Merging.  Thanks @almaleksia !

**Comment by almaleksia** [comment]
@mrocklin thank you :)

**Comment by mrocklin** [comment]
This is one of the nicer first PRs I've seen.  

If there is anything we can do to keep you busy here please don't hesitate to speak up :)

**Comment by almaleksia** [comment]
@mrocklin thank you for your feedback, I really appreciate it. I'll be happy to contribute further as soon as I find another issue that I could do. If you have something on mind I can take it :)

**Comment by mrocklin** [comment]
Happy to point you to good issues.  What are your interests?  Just
dataframes?  Also arrays?  Machine learning algorithms?  Network
communication and process handling?

**Comment by almaleksia** [comment]
@mrocklin I'm not picky but smth related to ML algos would be great!

**Comment by TomAugspurger** [comment]
@almaleksia you might have a look at the dask-ml repository then: https://github.com/dask/dask-ml

My short-term wishlist includes

- dask-aware versions of scikit-learn's train_test_split / CV things (started at https://github.com/tomaugspurger/dask-ml/tree/train-test-split)
- collaborative filtering / recommendation systems: https://github.com/dask/dask-ml/issues/139, because it'd be fun
- High Cardinality Encoders: https://github.com/dask/dask-ml/issues/120
- Anything you're interested working on :)

**Comment by almaleksia** [comment]
@TomAugspurger thanks a lot, I‘ll take a look :)

**Comment by mrocklin** [comment]
Some additional Dask issues that may be interesting.  They are either dataframe related, or are good introductions to other topics.

-  https://github.com/dask/dask/issues/3376
- https://github.com/dask/dask/issues/3361 (I haven't looked deeply at this, but perhaps someone should)
- https://github.com/dask/dask/issues/3336
- https://github.com/dask/dask/issues/3333 (this gets into join code, which is a bit intense)
- https://github.com/dask/dask/issues/3302 (a good introduction into dask arrays, which are, I think, somewhat more interesting algorithmically) 
-  https://github.com/dask/dask/issues/3269
- https://github.com/dask/dask/issues/3253 (an introduction into the local scheduler, this would result in an project that is external to the main codebase, living somewhere something like https://github.com/dask/dask-celery
-  https://github.com/dask/dask/issues/3241 (some more algorithmic and mathematical ideas)
-  https://github.com/dask/dask/issues/3240 (this connects to new work in pandas, cyberpandas, geopandas.  it would be very nice to see dask.dataframe adopt this as well)
-  https://github.com/dask/dask/issues/3206

**Comment by almaleksia** [comment]
@mrocklin great list, thanks!
